source/extensions/common/proxy_protocol/proxy_protocol_header.cc

```diff
@@ -111,19 +111,47 @@ void generateV2Header(const Network::Address::Ip& source_address,
 }
 
 bool generateV2Header(const Network::ProxyProtocolData& proxy_proto_data, Buffer::Instance& out,
-                      bool pass_all_tlvs, const absl::flat_hash_set<uint8_t>& pass_through_tlvs) {
-  uint64_t extension_length = 0;
-  for (auto&& tlv : proxy_proto_data.tlv_vector_) {
+                      bool pass_all_tlvs, const absl::flat_hash_set<uint8_t>& pass_through_tlvs,
+                      const std::vector<Envoy::Network::ProxyProtocolTLV>& custom_tlvs) {
+  std::vector<Envoy::Network::ProxyProtocolTLV> combined_tlv_vector;
+  std::vector<Envoy::Network::ProxyProtocolTLV> final_tlvs;
+  combined_tlv_vector.reserve(custom_tlvs.size() + proxy_proto_data.tlv_vector_.size());
+
+  absl::flat_hash_set<uint8_t> seen_types;
+  for (const auto& tlv : custom_tlvs) {
+    ASSERT(!seen_types.contains(tlv.type));
+    combined_tlv_vector.emplace_back(tlv);
+    seen_types.insert(tlv.type);
+  }
+
+  // Combine TLVs from the proxy_proto_data with the custom TLVs.
+  for (const auto& tlv : proxy_proto_data.tlv_vector_) {
     if (!pass_all_tlvs && !pass_through_tlvs.contains(tlv.type)) {
+      // Skip any TLV that is not in the set of passthrough TLVs.
       continue;
     }
-    extension_length += PROXY_PROTO_V2_TLV_TYPE_LENGTH_LEN + tlv.value.size();
-    if (extension_length > std::numeric_limits<uint16_t>::max()) {
-      ENVOY_LOG_MISC(
-          warn, "Generating Proxy Protocol V2 header: TLVs exceed length limit {}, already got {}",
-          std::numeric_limits<uint16_t>::max(), extension_length);
-      return false;
+    if (seen_types.contains(tlv.type)) {
+      // Skip any duplicate TLVs from being added to the combined TLV vector.
+      ENVOY_LOG_EVERY_POW_2_MISC(info, "Skipping duplicate TLV type {}", tlv.type);
+      continue;
     }
+    seen_types.insert(tlv.type);
+    combined_tlv_vector.emplace_back(tlv);
+  }
+
+  // Filter out TLVs that would exceed the 65535 limit.
+  uint64_t extension_length = 0;
+  bool skipped_tlvs = false;
+  for (auto&& tlv : combined_tlv_vector) {
+    uint64_t new_size = extension_length + PROXY_PROTO_V2_TLV_TYPE_LENGTH_LEN + tlv.value.size();
+    if (new_size > std::numeric_limits<uint16_t>::max()) {
+      ENVOY_LOG_MISC(warn, "Skipping TLV type {} because adding it would exceed the 65535 limit.",
+                     tlv.type);
+      skipped_tlvs = true;
+      continue;
+    }
+    extension_length = new_size;
+    final_tlvs.push_back(tlv);
   }
```

Comment:
Truncated custom or passthrough TLVs that exceed the max length in https://github.com/envoyproxy/envoy/pull/37591/commits/ad46d237feaee356618a904f327601d8f5efac1d. There were some header unit tests that were failing as a result as https://github.com/envoyproxy/envoy/blob/3122ab3a7dff1bcaf6c26d86f3425180311ca2c7/source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc#L106-L109 expects this function to return false to increment the stats.

I went back and forth on whether returning false here would be ambiguous when writing the output header was successful, but I didn't want to extend the function signature to return (bool written, bool skipped_tlvs) to handle this edge case at the call sites.